keps/sig-node/2371-cri-pod-container-stats/README.md

```diff
@@ -0,0 +1,730 @@
+2. For the Beta release, evaluate the future of the `/metrics/cadvisor` endpoint. We can either:
+    - Deprecate the endpoint, and extend the Summary API to report the required fields.
+    - Move the reporter of `/metrics/cadvisor` from cAdvisor to the CRI implementation.
+    - Extend the CRI and have the CRI implementation report the required stats, and have the Kubelet implement the `/metrics/cadvisor` endpoint. 
+3. For the GA release, the CRI implementation should be the source of truth for all pod and container level metrics that external parties rely on (no matter how many endpoints the Kubelet advertises).
+
+
+### Risks and Mitigations
+
+- To properly move to CRI stats, it is likely there are some metrics that we'll want to not support (Accelerator/UserDefined). We should be careful to not break entities that rely on these metrics.
+- A large part of this work is changing the source of the Summary API metrics. In doing so, there is a risk that collecting from a new source will change how the stats look in aggregate (and risk bugs popping up in new areas).
+- cAdvisor has a long history of collecting these stats. There is a risk that changing the source of the stats to the CRI implementation can cause performance regressions
+as cAdvisor is fine tuned to perform in an adequate manner.
+    - CRI implementations should do performance regression analyses to ensure the change does not regress too much.
+
+## Design Details
+
+
+#### CRI Implementation
+The CRI implementation will need to be extended to support reporting the full set of container-level from the [Summary API][#summary-container-stats-object].
+
+##### ContainerStats additions
+Currently, the CRI endpoints `{,List}ContainerStats` report the following fields for each container:
+- CPU
+    - `usage_core_nano_seconds`
+- Memory
+    - `working_set_bytes`
+- Filesystem
+    - `inodes_used`
+    - `used_bytes`
+
+These correspond to some fields of the [ContainerStats][#summary-container-stats-object] object reported by cAdvisor. Each of the CRI Stats fields will have to be extended to support the remaining fields. Here are some concrete additions the CRI will need, as well as notes showing what fields correspond to what between CRI and cAdvisor (note: a table version of this can be seen above):
+```
+-// CpuUsage provides the CPU usage information.
+-message CpuUsage {
+-    // Timestamp in nanoseconds at which the information were collected. Must be > 0.
++    // Corresponds to Stats Summary API CPUStats Time field
+-    int64 timestamp = 1;
+-    // Cumulative CPU usage (sum across all cores) since object creation.
++    // Corresponds to Stats Summary API CPUStats UsageCoreNanoSeconds
+-    UInt64Value usage_core_nano_seconds = 2;
++    // Total CPU usage (sum of all cores) averaged over the sample window.
++    UInt64Value usage_nano_seconds = 3;
+-}
+
+-// MemoryUsage provides the memory usage information.
+-message MemoryUsage {
```

Comment:
i am not a Windows expert, but I am curious if all these fields have the same meaning on Windows platform.

its possible we would want a struct per platform LinuxContainerStats or WindowsContainerStats for example.

Reply:
I know that at least with Docker CPU usage is represented differently for Windows vs Linux (this has actually led to several bugs in dockershim where conversions were done in some places and not others)
I forget if this was changed for contianerD on Windows or not and need to look this up.

Reply:
I am not sure how these changes could be introduced in a backwards compatible way. Would we do away with the current struct, and introduce a new one like {Linux,Windows}MemoryUsage?

Reply:
I've added a field in an Open Questions section for this